gomap/gomap.gobra

```diff
@@ -0,0 +1,163 @@
+decreases
+pure func IsSubset(m1, m2 map[int]int) (result bool) {
+	return domain(m1) subset domain(m2) &&
+			forall k int :: {IsEqualOnKey(m1, m2, k)} {k in domain(m1)} (k in domain(m1)) ==>
+				IsEqualOnKey(m1, m2, k)
+}
+
+// Maps are disjoint iff their domains are disjoint.
+ghost
+requires acc(m1, _)
+requires acc(m2, _)
+ensures result == dicts.AreDisjoint(ToDict(m1), ToDict(m2))
+decreases
+pure func AreDisjoint(m1, m2 map[int]int) (result bool) {
+	return sets.AreDisjoint(domain(m1), domain(m2))
+}
+
+ghost
+requires acc(m, _)
+ensures result == dicts.IsInjective(ToDict(m))
+opaque
+decreases
+pure func IsInjective(m map[int]int) (result bool) {
+	return let _ := reveal dicts.IsInjective(ToDict(m)) in
+		forall k1, k2 int :: {m[k1], m[k2]} (k1 != k2 && k1 in domain(m) && k2 in domain(m)) ==>
+			m[k1] != m[k2]
+}
+
+ghost
+requires acc(m, _)
+ensures result == dicts.IsMonotonic(ToDict(m))
+opaque
+decreases
+pure func IsMonotonic(m map[int]int) (result bool) {
+	return let _ := reveal dicts.IsMonotonic(ToDict(m)) in
+		forall k1, k2 int :: {m[k1], m[k2]} (k1 in domain(m) && k2 in domain(m) && k1 <= k2) ==>
+			m[k1] <= m[k2]
+}
+
+// True iff a map is monotonic in the range [start, +∞)
+ghost
+requires acc(m, _)
+ensures result == dicts.IsMonotonicFrom(ToDict(m), start)
+opaque
+decreases
+pure func IsMonotonicFrom(m map[int]int, start int) (result bool) {
+	return let _ := reveal dicts.IsMonotonicFrom(ToDict(m), start) in
+		forall k1, k2 int :: {m[k1], m[k2]} (k1 in domain(m) && k2 in domain(m) && start <= k1 && k1 <= k2) ==>
+			m[k1] <= m[k2]
+}
+
+// True iff a map is monotonic in the interval [start, end).
+ghost
+requires acc(m, _)
+ensures result == dicts.IsMonotonicFromTo(ToDict(m), start, end)
+opaque
+decreases
+pure func IsMonotonicFromTo(m map[int]int, start, end int) (result bool) {
+	return let _ := reveal dicts.IsMonotonicFromTo(ToDict(m), start, end) in
+		forall k1, k2 int :: {m[k1], m[k2]} (k1 in domain(m) &&
+						     k2 in domain(m) &&
+		 				     start <= k1 && k1 <= k2 && k2 < end) ==>
+							m[k1] <= m[k2]
+}
+
+// True iff two maps are equal in the interval [start, end).
+ghost
+requires acc(m1, _)
+requires acc(m2, _)
+ensures result == dicts.IsEqualInRange(ToDict(m1), ToDict(m2), start, end)
+opaque
+decreases
+pure func IsEqualInRange(m1, m2 map[int]int, start, end int) (result bool) {
+	return let _ := reveal dicts.IsEqualInRange(ToDict(m1), ToDict(m2), start, end) in
+		(forall k int :: {m1[k], m2[k]} {k in domain(m1)} {k in domain(m2)} (start <= k && k < end) ==>
+			k in domain(m1) && k in domain(m2) && m1[k] == m2[k])
+}
+
+// True iff m1 and m2 agree on all keys that their domains share.
+ghost
+requires acc(m1, _)
+requires acc(m2, _)
+ensures result == dicts.Agree(ToDict(m1), ToDict(m2))
+decreases
+pure func Agree(m1, m2 map[int]int) (result bool) {
+	return forall k int :: {m1[k], m2[k]} k in (domain(m1) intersection domain(m2)) ==>
+			m1[k] == m2[k]
```

Comment:
identation looks inconsistent

Reply:
I agree that it looks off; however, it is consistent in the sense that in `forall x :: a ==> b`, if `b` is in a new line, it is indented by one relative to `forall` (compare with `IsMonotonic` and `IsInjective`, for example) . Removing that indentation everywhere probably will look off again somewhere.

Reply:
I think that, in some places, the indentation is relative to the the `return`. At any rate, let's ignore it for now. Hopefully, we will have an automated formatter at some point